certora/specs/ERC20Invariants.spec

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+    function upgradeToAndCall(address, bytes) external => NONDET DELETE;
+}
+
+// Paramater for any address that is not the zero address.
+persistent ghost address A {
+    axiom A != 0;
+}
+
+// Partial sum of delegated votes to parameterized address A.
```

Comment:
why do we need partial sums?

Reply:
This is the point of the new "trick" to be able to prove things about the total supply and particular account, for example that the sum of balances of 2 accounts is lower or equal to the total supply.

The idea is to be able to distinguish updates from different accounts. This way this `sumOfVotes` ghost variables does not blend all updates into a big total sum. Then you can talk about specific balances, such as in the invariant `sumOfBalancesGrowsCorrectly`